slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/interchange/InterchangeV5Importer.java

```diff
@@ -0,0 +1,188 @@
+    importAttestations(h, validator, signedAttestationNode);
+  }
+
+  private void importBlocks(
+      final Handle h, final Validator validator, final ArrayNode signedBlocksNode)
+      throws JsonProcessingException {
+    for (int i = 0; i < signedBlocksNode.size(); i++) {
+      final SignedBlock jsonBlock = mapper.treeToValue(signedBlocksNode.get(i), SignedBlock.class);
+      final BlockValidator blockValidator =
+          new BlockValidator(
+              h,
+              jsonBlock.getSigningRoot(),
+              jsonBlock.getSlot(),
+              validator.getId(),
+              signedBlocksDao);
+
+      if (blockValidator.existsInDatabase()) {
+        LOG.debug(
+            "Block {} for validator {} already exists in database, not imported",
+            i,
+            validator.getPublicKey());
+        continue; // DO NOT duplicate import.
+      } else if (blockValidator.directlyConflictsWithExistingEntry()) {
+        LOG.debug(
+            "Block {} for validator {} conflicts with on slot {} in database",
+            i,
+            validator.getPublicKey(),
+            jsonBlock.getSlot());
+      }
+
+      signedBlocksDao.insertBlockProposal(
+          h,
+          new tech.pegasys.web3signer.slashingprotection.dao.SignedBlock(
+              validator.getId(), jsonBlock.getSlot(), jsonBlock.getSigningRoot()));
+    }
+  }
+
+  private void importAttestations(
+      final Handle h, final Validator validator, final ArrayNode signedAttestationNode)
+      throws JsonProcessingException {
+    for (int i = 0; i < signedAttestationNode.size(); i++) {
+      final SignedAttestation jsonAttestation =
+          mapper.treeToValue(signedAttestationNode.get(i), SignedAttestation.class);
+      final AttestationValidator attestationValidator =
+          new AttestationValidator(
+              h,
+              validator.getPublicKey(),
+              jsonAttestation.getSigningRoot(),
+              jsonAttestation.getSourceEpoch(),
+              jsonAttestation.getTargetEpoch(),
+              validator.getId(),
+              signedAttestationsDao);
+
+      if (attestationValidator.sourceGreaterThanTargetEpoch()) {
+        // exit
+        throw new IllegalArgumentException(
+            String.format(
+                "Attestation #%d for validator %s - source is great than target epoch",
+                i, validator.getPublicKey()));
+      } else if (attestationValidator.existsInDatabase()) {
+        continue; // do not re-insert
+      } else if (attestationValidator.isSurroundedByExistingAttestation()) {
+        LOG.warn(
+            "Attestation {} of validator {} is surrounded by existing entries",
+            i,
+            validator.getPublicKey());
+      } else if (attestationValidator.surroundsExistingAttestation()) {
+        LOG.warn(
+            "Attestation {} of validator {} surrounds an existing entry",
+            i,
+            validator.getPublicKey());
+      }
```

Comment:
Why do these checks need to be done? I thought we were importing the data as is

Reply:
they're just for warnings, not preventing importing - only source<target is a true failure.